src/main/java/core/basesyntax/MyLinkedList.java

```diff
@@ -3,45 +3,172 @@
-        return null;
+        indexChecker(index);
+        return getNodeByIndex(index).value;
     }
 
     @Override
     public T set(T value, int index) {
-        return null;
+        indexChecker(index);
+        Node<T> current = getNodeByIndex(index);
+        T removedValue = current.value;
+        current.value = value;
+        return removedValue;
     }
 
     @Override
     public T remove(int index) {
-        return null;
+        indexChecker(index);
+        Node<T> current = getNodeByIndex(index);
+        unlink(current);
+        return current.value;
     }
 
     @Override
     public boolean remove(T object) {
+        Node<T> current = first;
+        while (current != null) {
+            if ((current.value == null && object == null)
+                    || (current.value != null && (current.value.equals(object)))) {
+                unlink(current);
+                return true;
+            }
+            current = current.nextElement;
+        }
         return false;
     }
 
     @Override
     public int size() {
-        return 0;
+        return size;
     }
 
     @Override
     public boolean isEmpty() {
-        return false;
+        return size() == DEFAULT_INDEX;
+    }
+
+    private void indexChecker(int index) {
+        if (index < DEFAULT_INDEX || index >= size()) {
+            throw new IndexOutOfBoundsException("Index is out of bounds for size "
+                    + size() + " with element " + index);
+        }
+    }
+
+    private void indexCheckerForAdding(int index) {
+        if (index < DEFAULT_INDEX || index > size()) {
+            throw new IndexOutOfBoundsException("Index is out of bounds for size "
+                    + size() + " with element " + index);
+        }
+    }
+
+    private Node<T> getNodeByIndex(int index) {
+        if (index <= size() / DIVIDER) {
+            return countFromFirst(index);
+        } else {
+            return countFromLast(index);
+        }
+    }
+
+    private Node<T> countFromFirst(int index) {
+        int counter = DEFAULT_INDEX;
+        Node<T> current = first;
+        while (counter < index) {
+            current = current.nextElement;
+            counter++;
+        }
+        return current;
+    }
+
+    private Node<T> countFromLast(int index) {
+        int counter = size() - 1;
+        Node<T> current = last;
+        while (counter > index) {
+            current = current.prevElement;
+            counter--;
+        }
+        return current;
+    }
+
+    private void unlink(Node<T> removedNode) {
+        Node<T> prevNode = removedNode.prevElement;
+        Node<T> nextNode = removedNode.nextElement;
+        if (prevNode == null) {
+            first = nextNode;
+        } else {
+            prevNode.nextElement = nextNode;
+        }
+        if (nextNode == null) {
+            last = prevNode;
+        } else {
+            nextNode.prevElement = prevNode;
+        }
+        size--;
+    }
```

Comment:
![image](https://user-images.githubusercontent.com/58032907/195824727-3e470769-03b8-4376-b317-23a9e687b538.png)
after 153 line we should write `nextNode.prevElement = null` and similar for other cases